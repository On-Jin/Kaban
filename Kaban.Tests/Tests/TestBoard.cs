using System.Net;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using FluentAssertions;
using Kaban.Data;
using Kaban.GraphQL.Boards;
using Kaban.Tests.Setup;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit.Abstractions;

namespace Kaban.Tests.Tests;

[Collection(nameof(WebAppCollectionFixture))]
public class TestBoard
{
    private readonly ITestOutputHelper _testOutputHelper;
    private readonly JsonSerializerOptions _jsonOptionInputGraphQl;
    private readonly HttpClient _httpClientDiscord;
    private readonly HttpClient _httpClientShadow;

    private readonly Func<Task> _resetDatabase;
    private readonly WebAppFactory _factory;

    public TestBoard(WebAppFactory factory, ITestOutputHelper testOutputHelper)
    {
        _testOutputHelper = testOutputHelper;
        _jsonOptionInputGraphQl = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
        };
        _jsonOptionInputGraphQl.Converters.Add(new JsonStringEnumConverter());


        _factory = factory;
        _testOutputHelper = testOutputHelper;
        _resetDatabase = factory.ResetDatabaseAsync;

        _httpClientDiscord = factory.CreateClient();
        _httpClientDiscord.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue(TestAuthHandler.AuthenticationScheme);
        _httpClientShadow = factory.CreateClient();

        _httpClientShadow.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue(TestAuthHandler.AuthenticationScheme);
        _httpClientShadow.DefaultRequestHeaders.Add(TestHelper.HeaderUserGuid, TestHelper.GuidShadowAuth);
    }

    private async Task<HttpResponseMessage> MakeGraphQL_Request(HttpClient httpClient, string pathToGqlQueryFile,
        string input = "", string? mutationName = null)
    {
        string jsonContent =
            await Utilities.FileGqlToJsonQuery(pathToGqlQueryFile, input.Replace("\\\"", "\""), mutationName);
        _testOutputHelper.WriteLine($"\\\\\\\\ Input");
        _testOutputHelper.WriteLine(jsonContent);
        _testOutputHelper.WriteLine($"\\\\\\\\");
        StringContent httpContent = new(jsonContent, System.Text.Encoding.UTF8, "application/json");
        HttpResponseMessage response = await httpClient.PostAsync("/graphql", httpContent);

        _testOutputHelper.WriteLine($"//// Response from Query from {pathToGqlQueryFile}");
        _testOutputHelper.WriteLine(await response.Content.ReadAsStringAsync());
        _testOutputHelper.WriteLine("////");

        return response;
    }

    private async Task<HttpResponseMessage> QueryBoards(HttpClient httpClient)
    {
        HttpResponseMessage response = await MakeGraphQL_Request(httpClient,
            "Resources/Ql_Query_Boards.gql");
        response.StatusCode.Should().Be(HttpStatusCode.OK);

        return response;
    }

    private async Task<HttpResponseMessage> AddBoard(HttpClient httpClient, AddBoardInput? addBoardInput = null)
    {
        addBoardInput ??= new AddBoardInput("Example");

        HttpResponseMessage response = await MakeGraphQL_Request(httpClient,
            "Resources/Ql_AddBoard.gql",
            JsonSerializer.Serialize(addBoardInput, _jsonOptionInputGraphQl));
        response.StatusCode.Should().Be(HttpStatusCode.OK);

        return response;
    }

    private async Task<HttpResponseMessage> DeleteBoard(HttpClient httpClient,
        DeleteBoardInput? deleteBoardInput = null)
    {
        deleteBoardInput ??= new DeleteBoardInput(1);

        HttpResponseMessage response = await MakeGraphQL_Request(httpClient,
            "Resources/Ql_DeleteBoard.gql",
            JsonSerializer.Serialize(deleteBoardInput, _jsonOptionInputGraphQl));
        response.StatusCode.Should().Be(HttpStatusCode.OK);

        return response;
    }

    private async Task<HttpResponseMessage> PatchBoard(HttpClient httpClient,
        PatchBoardInput? patchBoardInput = null)
    {
        patchBoardInput ??= new PatchBoardInput(1, "New Name Example!");

        HttpResponseMessage response = await MakeGraphQL_Request(httpClient,
            "Resources/Ql_PatchBoard.gql",
            JsonSerializer.Serialize(patchBoardInput, _jsonOptionInputGraphQl));
        response.StatusCode.Should().Be(HttpStatusCode.OK);

        return response;
    }

    private async Task GenerateKabanBoardToShadowUser(AppDbContext db)
    {
        var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
        var boards = await GenerateHelper.GenerateDefaultKabanBoards();
        user.Boards.AddRange(boards);
        await db.SaveChangesAsync();
    }

    [Fact]
    public async Task GraphQL_AddBoard()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();
        await GenerateKabanBoardToShadowUser(db);

        HttpResponseMessage response = await AddBoard(_httpClientShadow);

        string jsonWrapped = await response.Content.ReadAsStringAsync();
        _testOutputHelper.WriteLine(jsonWrapped);

        jsonWrapped.MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_DeleteBoard()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();
        await GenerateKabanBoardToShadowUser(db);

        int id = 0;
        {
            var response = await AddBoard(_httpClientShadow);
            var jsonWrapped = await response.Content.ReadAsStringAsync();
            _testOutputHelper.WriteLine(jsonWrapped);
            var jsonObject = JsonNode.Parse(jsonWrapped)!.AsObject();
            id = int.Parse(jsonObject["data"]!["addBoard"]!["board"]["id"]!.ToString());
            _testOutputHelper.WriteLine($"{id}");
        }

        {
            {
                await DeleteBoard(_httpClientShadow, new DeleteBoardInput(id));
            }
            {
                var response = await QueryBoards(_httpClientShadow);
                string jsonWrapped = await response.Content.ReadAsStringAsync();
                JsonSerializer
                    .Serialize(JsonSerializer.Deserialize<object>(jsonWrapped), TestHelper.JsonSerializerOptions)
                    .MatchSnapshot();
            }
        }
    }


    [Fact]
    public async Task GraphQL_PatchBoard()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();
        await GenerateKabanBoardToShadowUser(db);

        var response = await PatchBoard(_httpClientShadow);
        JsonSerializer
            .Serialize(
                JsonSerializer.Deserialize<object>(await response.Content.ReadAsStringAsync()),
                TestHelper.JsonSerializerOptions)
            .MatchSnapshot();
    }
}