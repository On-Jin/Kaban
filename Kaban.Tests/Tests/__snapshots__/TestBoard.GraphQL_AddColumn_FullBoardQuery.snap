{
  "data": {
    "boards": [
      {
        "id": 1,
        "name": "Platform Launch",
        "columns": [
          {
            "id": 1,
            "name": "Todo",
            "mainTasks": [
              {
                "description": "",
                "id": 1,
                "status": "Todo",
                "title": "Build UI for onboarding flow",
                "subTasks": [
                  {
                    "id": 1,
                    "isCompleted": true,
                    "title": "Sign up page"
                  },
                  {
                    "id": 2,
                    "isCompleted": false,
                    "title": "Sign in page"
                  },
                  {
                    "id": 3,
                    "isCompleted": false,
                    "title": "Welcome page"
                  }
                ]
              },
              {
                "description": "",
                "id": 2,
                "status": "Todo",
                "title": "Build UI for search",
                "subTasks": [
                  {
                    "id": 4,
                    "isCompleted": false,
                    "title": "Search page"
                  }
                ]
              },
              {
                "description": "",
                "id": 3,
                "status": "Todo",
                "title": "Build settings UI",
                "subTasks": [
                  {
                    "id": 5,
                    "isCompleted": false,
                    "title": "Account page"
                  },
                  {
                    "id": 6,
                    "isCompleted": false,
                    "title": "Billing page"
                  }
                ]
              },
              {
                "description": "Once we feel version one is ready, we need to rigorously test it both internally and externally to identify any major gaps.",
                "id": 4,
                "status": "Todo",
                "title": "QA and test all major user journeys",
                "subTasks": [
                  {
                    "id": 7,
                    "isCompleted": false,
                    "title": "Internal testing"
                  },
                  {
                    "id": 8,
                    "isCompleted": false,
                    "title": "External testing"
                  }
                ]
              }
            ]
          },
          {
            "id": 2,
            "name": "Doing",
            "mainTasks": [
              {
                "description": "",
                "id": 5,
                "status": "Doing",
                "title": "Design settings and search pages",
                "subTasks": [
                  {
                    "id": 9,
                    "isCompleted": true,
                    "title": "Settings - Account page"
                  },
                  {
                    "id": 10,
                    "isCompleted": true,
                    "title": "Settings - Billing page"
                  },
                  {
                    "id": 11,
                    "isCompleted": false,
                    "title": "Search page"
                  }
                ]
              },
              {
                "description": "",
                "id": 6,
                "status": "Doing",
                "title": "Add account management endpoints",
                "subTasks": [
                  {
                    "id": 12,
                    "isCompleted": true,
                    "title": "Upgrade plan"
                  },
                  {
                    "id": 13,
                    "isCompleted": true,
                    "title": "Cancel plan"
                  },
                  {
                    "id": 14,
                    "isCompleted": false,
                    "title": "Update payment method"
                  }
                ]
              },
              {
                "description": "",
                "id": 7,
                "status": "Doing",
                "title": "Design onboarding flow",
                "subTasks": [
                  {
                    "id": 15,
                    "isCompleted": true,
                    "title": "Sign up page"
                  },
                  {
                    "id": 16,
                    "isCompleted": false,
                    "title": "Sign in page"
                  },
                  {
                    "id": 17,
                    "isCompleted": false,
                    "title": "Welcome page"
                  }
                ]
              },
              {
                "description": "",
                "id": 8,
                "status": "Doing",
                "title": "Add search enpoints",
                "subTasks": [
                  {
                    "id": 18,
                    "isCompleted": true,
                    "title": "Add search endpoint"
                  },
                  {
                    "id": 19,
                    "isCompleted": false,
                    "title": "Define search filters"
                  }
                ]
              },
              {
                "description": "",
                "id": 9,
                "status": "Doing",
                "title": "Add authentication endpoints",
                "subTasks": [
                  {
                    "id": 20,
                    "isCompleted": true,
                    "title": "Define user model"
                  },
                  {
                    "id": 21,
                    "isCompleted": false,
                    "title": "Add auth endpoints"
                  }
                ]
              },
              {
                "description": "We know what we\u0027re planning to build for version one. Now we need to finalise the first pricing model we\u0027ll use. Keep iterating the subtasks until we have a coherent proposition.",
                "id": 10,
                "status": "Doing",
                "title": "Research pricing points of various competitors and trial different business models",
                "subTasks": [
                  {
                    "id": 22,
                    "isCompleted": true,
                    "title": "Research competitor pricing and business models"
                  },
                  {
                    "id": 23,
                    "isCompleted": false,
                    "title": "Outline a business model that works for our solution"
                  },
                  {
                    "id": 24,
                    "isCompleted": false,
                    "title": "Talk to potential customers about our proposed solution and ask for fair price expectancy"
                  }
                ]
              }
            ]
          },
          {
            "id": 3,
            "name": "Done",
            "mainTasks": [
              {
                "description": "Ensure the layout continues to make sense and we have strong buy-in from potential users.",
                "id": 11,
                "status": "Done",
                "title": "Conduct 5 wireframe tests",
                "subTasks": [
                  {
                    "id": 25,
                    "isCompleted": true,
                    "title": "Complete 5 wireframe prototype tests"
                  }
                ]
              },
              {
                "description": "Create a greyscale clickable wireframe prototype to test our asssumptions so far.",
                "id": 12,
                "status": "Done",
                "title": "Create wireframe prototype",
                "subTasks": [
                  {
                    "id": 26,
                    "isCompleted": true,
                    "title": "Create clickable wireframe prototype in Balsamiq"
                  }
                ]
              },
              {
                "description": "Keep iterating through the subtasks until we\u0027re clear on the core concepts for the app.",
                "id": 13,
                "status": "Done",
                "title": "Review results of usability tests and iterate",
                "subTasks": [
                  {
                    "id": 27,
                    "isCompleted": true,
                    "title": "Meet to review notes from previous tests and plan changes"
                  },
                  {
                    "id": 28,
                    "isCompleted": true,
                    "title": "Make changes to paper prototypes"
                  },
                  {
                    "id": 29,
                    "isCompleted": true,
                    "title": "Conduct 5 usability tests"
                  }
                ]
              },
              {
                "description": "",
                "id": 14,
                "status": "Done",
                "title": "Create paper prototypes and conduct 10 usability tests with potential customers",
                "subTasks": [
                  {
                    "id": 30,
                    "isCompleted": true,
                    "title": "Create paper prototypes for version one"
                  },
                  {
                    "id": 31,
                    "isCompleted": true,
                    "title": "Complete 10 usability tests"
                  }
                ]
              },
              {
                "description": "We need to define and refine our core product. Interviews will help us learn common pain points and help us define the strongest MVP.",
                "id": 15,
                "status": "Done",
                "title": "Market discovery",
                "subTasks": [
                  {
                    "id": 32,
                    "isCompleted": true,
                    "title": "Interview 10 prospective customers"
                  }
                ]
              },
              {
                "description": "",
                "id": 16,
                "status": "Done",
                "title": "Competitor analysis",
                "subTasks": [
                  {
                    "id": 33,
                    "isCompleted": true,
                    "title": "Find direct and indirect competitors"
                  },
                  {
                    "id": 34,
                    "isCompleted": true,
                    "title": "SWOT analysis for each competitor"
                  }
                ]
              },
              {
                "description": "We need to get a solid overview of the market to ensure we have up-to-date estimates of market size and demand.",
                "id": 17,
                "status": "Done",
                "title": "Research the market",
                "subTasks": [
                  {
                    "id": 35,
                    "isCompleted": true,
                    "title": "Write up research analysis"
                  },
                  {
                    "id": 36,
                    "isCompleted": true,
                    "title": "Calculate TAM"
                  }
                ]
              }
            ]
          },
          {
            "id": 10,
            "name": "Example",
            "mainTasks": []
          }
        ]
      },
      {
        "id": 2,
        "name": "Marketing Plan",
        "columns": [
          {
            "id": 4,
            "name": "Todo",
            "mainTasks": [
              {
                "description": "",
                "id": 18,
                "status": "Todo",
                "title": "Plan Product Hunt launch",
                "subTasks": [
                  {
                    "id": 37,
                    "isCompleted": false,
                    "title": "Find hunter"
                  },
                  {
                    "id": 38,
                    "isCompleted": false,
                    "title": "Gather assets"
                  },
                  {
                    "id": 39,
                    "isCompleted": false,
                    "title": "Draft product page"
                  },
                  {
                    "id": 40,
                    "isCompleted": false,
                    "title": "Notify customers"
                  },
                  {
                    "id": 41,
                    "isCompleted": false,
                    "title": "Notify network"
                  },
                  {
                    "id": 42,
                    "isCompleted": false,
                    "title": "Launch!"
                  }
                ]
              },
              {
                "description": "",
                "id": 19,
                "status": "Todo",
                "title": "Share on Show HN",
                "subTasks": [
                  {
                    "id": 43,
                    "isCompleted": false,
                    "title": "Draft out HN post"
                  },
                  {
                    "id": 44,
                    "isCompleted": false,
                    "title": "Get feedback and refine"
                  },
                  {
                    "id": 45,
                    "isCompleted": false,
                    "title": "Publish post"
                  }
                ]
              },
              {
                "description": "",
                "id": 20,
                "status": "Todo",
                "title": "Write launch article to publish on multiple channels",
                "subTasks": [
                  {
                    "id": 46,
                    "isCompleted": false,
                    "title": "Write article"
                  },
                  {
                    "id": 47,
                    "isCompleted": false,
                    "title": "Publish on LinkedIn"
                  },
                  {
                    "id": 48,
                    "isCompleted": false,
                    "title": "Publish on Inndie Hackers"
                  },
                  {
                    "id": 49,
                    "isCompleted": false,
                    "title": "Publish on Medium"
                  }
                ]
              }
            ]
          },
          {
            "id": 5,
            "name": "Doing",
            "mainTasks": []
          },
          {
            "id": 6,
            "name": "Done",
            "mainTasks": []
          }
        ]
      },
      {
        "id": 3,
        "name": "Roadmap",
        "columns": [
          {
            "id": 7,
            "name": "Now",
            "mainTasks": [
              {
                "description": "",
                "id": 21,
                "status": "Now",
                "title": "Launch version one",
                "subTasks": [
                  {
                    "id": 50,
                    "isCompleted": false,
                    "title": "Launch privately to our waitlist"
                  },
                  {
                    "id": 51,
                    "isCompleted": false,
                    "title": "Launch publicly on PH, HN, etc."
                  }
                ]
              },
              {
                "description": "Beyond the initial launch, we\u0027re keeping the initial roadmap completely empty. This meeting will help us plan out our next steps based on actual customer feedback.",
                "id": 22,
                "status": "Now",
                "title": "Review early feedback and plan next steps for roadmap",
                "subTasks": [
                  {
                    "id": 52,
                    "isCompleted": false,
                    "title": "Interview 10 customers"
                  },
                  {
                    "id": 53,
                    "isCompleted": false,
                    "title": "Review common customer pain points and suggestions"
                  },
                  {
                    "id": 54,
                    "isCompleted": false,
                    "title": "Outline next steps for our roadmap"
                  }
                ]
              }
            ]
          },
          {
            "id": 8,
            "name": "Next",
            "mainTasks": []
          },
          {
            "id": 9,
            "name": "Later",
            "mainTasks": []
          }
        ]
      }
    ]
  }
}
